Mapping project page

bmegli opened the pull request:
Without youtube link yet. Video "in production".

The only controversial part of this pull should be x3d support (I hope).

Locally I have tested it working in:
-Firefox
-Chrome
-Opera
-Edge

All work. I had to fix the size of x3d container for Edge, it didn't scale well otherwise so it's fixed to 640x480 for now. Side note - Opera and Chrome seem to have the best performance with x3d (when rotating).  

Side note - the reconstruced face  is of my friend from work, not mine. He agreed to share the model.

dlech commented on projects/_posts/2016-08-07-Mapping.md:
```diff
@@ -0,0 +1,198 @@
+-   [How to interface XV11 LIDAR to EV3 using ev3dev]
+-   [EV3 Gyro vs CruizCore XG1300L vs Odometry - Position Estimation]
+
+{% include screenshot.html source="\images\projects\2016-08-07-Mapping\robot-phaseI.jpg" caption="Hardware used for testing in Phase I" %}
```
This is not windows :wink:

`/`, not `\`

There are other places this needs to be fixed as well.

bmegli commented:
Ok, the path separators corrected ;-)

Now I feel as if I have made a faux pas ;-)

bmegli commented:
Corrected incorrect link

bmegli commented:
One more thing.

I initially chose the name for the repository somewhat unfortunately.

It's not like I want to force it to ev3dev repositories, it was just mapping with ev3dev.

if ev3dev-mapping name for repository is a problem I can phase the repository to some different name.

WasabiFan commented:
I just pulled your changes to take a look, and it looks good to me. I really like that you were able to get the 3D models included! Could you bold the text "You can rotate, move and zoom with the mouse" so that people just skimming notice that the shiny things actually move too?

By the way, the "excerpt" field can be longer if you'd like it to be -- you aren't restricted on length. 

The only problem I noticed is that, because there is no video attached to the post to get a thumbnail from, this project looks a bit out-of-place on the index page. There isn't a workaround yet (the thumbnail logic relies on a YouTube video currently), but I'll work on a quick solution. This isn't a blocking issue for this PR as I see it.

WasabiFan commented:
@bmegli If you pull from master now (I'm assuming you know how to do that), I've added the ability to specify a project thumbnail manually. You can now set it in the front matter:

```yaml
thumbnail_override: "/images/projects/my-project/my-image.png"
```

bmegli commented:
@WasabiFan Thanks. I was actually wondering how to get the thumbnail.

> I really like that you were able to get the 3D models included!

Yes! I am also happy ;-)

It's still not perfect. While panning with middle mouse button pressed it also moves the scroolbars of the browser. Also I haven't managed to get the auto-scale for x3d container working properly in all (major) browsers so it's fixed size now.

The funny things is that we could go even further and embedd Unity web build of whole project on the page with replay of UDP communication. But this would need writing platform specific code for the web build. I don't think Unity web builds support threading.

I think we can hold merging this pull for a while, until I finish the youtube video.

Can I assume the fixed path  for linking back from youtube (non-editable after upload) to be:
`http://www.ev3dev.org/projects/2016/08/07/Mapping/`

WasabiFan commented:
> While panning with middle mouse button pressed it also moves the scroolbars of the browser. 

In Edge at least, I saw this happen once but trying again worked without scrolling. I suspect that something similar would happen in other browsers (once you focus on the scene, the controls will work correctly).

> Also I haven't managed to get the auto-scale for x3d container working properly in all (major) browsers so it's fixed size now.

Does it work to put the scene in a parent that is set to a percentage width? That might work.

> I think we can hold merging this pull for a while, until I finish the youtube video.

OK, that's fine.

> Can I assume the fixed path for linking back from youtube (non-editable after upload) to be:
http://www.ev3dev.org/projects/2016/08/07/Mapping/

In general, yes; however note that GH Pages' server is case-sensitive, so you need to make sure that casing is correct. It might make more sense to use a URL shortening service though, because with some services you can edit the destination of the link post-creation. That way, if the link isn't right or the page moves for whatever reason, you can fairly easily change it.

bmegli commented:
@WasabiFan 
> Does it work to put the scene in a parent that is set to a percentage width? That might work.
I think that I have tried with width, but I don't remember what I did with height. I will investigate. Thanks for the suggestion.

> It might make more sense to use a URL shortening service though
Huh, I do like the ev3dev.org address too much to spoil it with shortening.

bmegli commented:
I would like to to acknowledge all ev3dev members (and community) at the end of video and in video description (this means that it would be indexed by the browsers).

@dlech, @rhempel, @WasabiFan, @kortschak , @ddemidov, @fdetro, @harmooni, @bmegli:

Does anybody have something against?
Does anybody have something against using his/her real world name along the github pseudo in acknoledgments?
@fdetro, do you have something against affiliating your name with ev3dev?
@harmooni, are you ev3dev member? I just don't know your work, and I try too track what everybody does more or less.

Is there somebody else that should be acknowledged for the work on ev3dev who's not a member (apart from "The Community")?

bmegli commented:
> Does anybody have something against?
> Does anybody have something against using his/her real world name along the github pseudo in acknoledgments?

I am ok with using my github pseudo. I haven't decided yet for the real name. I have always published videos anonymously so far. So github pseudo - yes. Real name - maybe.

dlech commented:
> Is there somebody else that should be acknowledged for the work on ev3dev who's not a member (apart from "The Community")?

The ev3dev "members" are just people I have added so that I can assign issues to them. There are plenty of other people who have made just as significant contributions to ev3dev and deserve to be on this list as well. I would have a hard time coming up with a full list as there are many. It may be more fair to just acknowledge the ev3dev organization as a whole.

bmegli commented:
Ok! Sounds fair to me.

This is what I always do already anyway (acknowledging ev3dev).

fdetro commented:
Fine with me.

bmegli commented:
Done!

From my side this can be merged now.

kortschak commented on projects/_posts/2016-08-07-Mapping.md:
```diff
@@ -0,0 +1,253 @@
+
+project_homepage_url: "https://github.com/bmegli/ev3dev-mapping" 
+source_code_url: "https://github.com/bmegli/ev3dev-mapping" 
+#building_instructions_url: "" # how to build the model out of LEGO (*not* how to build the source code)
```
Is this a TODO?

bmegli replied:
@kortschak Thanks.  I left it  commented out in case I want it in the future. I am not giving building instructions. I am using various hardware to test this project depending on stage. There are some images of models used for testing on the project page.

kortschak replied:
No worries. As an aside, did you print the deck plate that you use in the phaseII model or is it available somewhere?

bmegli replied:
Oh, the MinuteBot technic baseplate?

http://www.minutebot.com/minuteproducts/base/

It's great to get some technic robot quickly up and running or for stationary models.

bmegli replied:
Probably I should add it in Building Instructions under the photos, you're right.

WasabiFan commented:
Aside from the empty `building_instructions_url` field that @kortschak mentioned (which I think should be filled in, given that there's an equivalent link in the text), this looks good to me.

bmegli commented:
@WasabiFan building_instructoins_url was left commented out in case I want to use it in the future.

There are no strict building instructions for this project because I am using it with various hardware. On the software side the orientation and position of the lidars is configurable with coordinates.

But you probably mean that I can link building instructions to Building Instructoins section with "model used for testing in phase II". Hmm, probably.

bmegli commented:
Oh, the building instructions link mentioned pertains to building (and running) the code ;-)

bmegli commented:
Ok, I have removed the confusing line with building instructions on github.

I leave the commented out building instructions url in case I want to use it in the future.

bmegli commented:
Ok, removed also the url after thinking it over.

WasabiFan commented:
> Oh, the building instructions link mentioned pertains to building (and running) the code ;-)

Ah, I see. In that case, I don't have a problem with there not being a URL there if there isn't something logical to link to.

bmegli commented:
Hmm, I unlisted the video from youtube. The sound levels are different across the machines I have tried it on to the point where voice can't be heard on some machines. I will upload new version.

I think that the project page can be merged with later only update of youtube link.

bmegli commented:
Ok, corrected sound amplitudes, reencoded with H.264, AAC-LC for sound, and uploaded to youtube.

The link from project page is also updated.

This material is ready now. I do feel tired...

bmegli commented:
Added information suggested by @kortschak , about MinuteBot technic base plate. I forgot this is not standard lego part.

Also key parts list in building instructions, this information was spread through the page.

bmegli commented:
Some links to keep homogenous structure for future phases

bmegli commented:
Waaaaaait, I think there is one more important thing to be done.

I had at some point 2 or 3 big binary files in my repository (uncompressed 3d models). They are probably in git history. If I don't cleanse the repository they will be downloaded whenever somebody clones the repository, and downloaded completely pointlessly (just as history of the repository).

WasabiFan commented:
> If I don't cleanse the repository they will be downloaded whenever somebody clones the repository, and downloaded completely pointlessly (just as history of the repository). 

Does GitHub do that? I've never tried, but I would imagine that it wouldn't preserve commits/blobs that aren't involved in the pull request branch when merging. i.e., your fork would be large, but the main repo wouldn't be.

bmegli commented:
@WasabiFan if I understand it correctly it does. Merging keeps the original history as it was developed. It is usually what is desired, not with big files however.

If I stupidly commited 4 MB file then overwritten it in other commit with compressed version both will stay in history.

I may be wrong natually, I am still learning. But I am going to take the safe path.

Ok, what I am going to do now:
- close this pull
- cleanup my repository to ev3dev github.io master to get rid of all the binary files versions history
- split this merge into 3 clean small branches pull-requests created anew: 
   -   xv11 tutorial minor updates
   -   x3d generic support
   -   Mapping Project 
- readup 

Even if unecessary it is safe.

WasabiFan commented:
> If I stupidly commited 4 MB file then overwritten it in other commit with compressed version both will stay in history.

Sorry, I think I was operating under an assumption that I didn't make clear. I was planning to have GitHub squash the commits from this PR into a single one when I/we merged it, which would mean that the blob containing the large file would still be in git storage, but there wouldn't be any commits referencing it. Either way, it sounds like you are going to do that yourself locally and open a new PR.

bmegli commented:
@WasabiFan 

> it sounds like you are going to do that yourself locally and open a new PR.

After 7 chapters and nearly 300 pages of Pro Git... ... the answer is yes 😃 

I just did it outside of git which was the way I understood at that time (branched, reset my repository with ev3dev and copied the final result files over it which was equivalent to squash...)

Anyway, `branch`, `total mess`, `pull`, `squash` workflow with possiblity to rewrite history is soooo awesome! The moment you notice that you can squash own pull on github with single click is enlightening.

WasabiFan commented:
Git is magic :sparkles:

The new squash feature really is nice. It also lets maintainers edit the commit message, so if the original contributor set the commit message to "Updated file.md", the maintainer can make it something more useful as they merge it. There's also an option for rebasing the new commits onto the base branch, which means that all the commits will be tacked onto the end of the target branch so they are all together. This works well when you did multiple things in the same PR but still want the "squash effect" and/or like them grouped together in the history.
